### PR TITLE
Add `core:container/rbtree`

### PR DESCRIPTION
**Pull request by Kelimion**
Add a Red-Black tree with configurable `$Key` and `$Value`,  as well as an `Iterator` and optional `on_remove` callback.
Also includes tests that verify it maintains Red-Black and Iterator invariants and doesn't leak.

Originally based on the CC0 implementation from literateprograms.org. But reworked to the same API used by @Yawning's excellent `core:container/avl` for ease of use.

### CODE REVIEW

File: core/container/rbtree/rbtree.odin [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+		case .Less:
+			n_ptr = &n._left
+		case .Greater:
+			n_ptr = &n._right
+		case .Equal:
+			return
+		}
+	}
+	_parent := n
+	n = new_clone(Node(Key, Value){key=key, value=value, _parent=_parent, _color=.Red}, t._node_allocator) // or_return
```
Comment (gingerBill) [inline core/container/rbtree/rbtree.odin:148]: Why not do?
```odin
err: runtime.Allocator_Error
n, err = new_clone(...)
if err != nil {
    return
}
```

File: core/container/rbtree/rbtree.odin [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+	}
+	free(node, t._node_allocator)
+	t._size -= 1
+	return true
+}
+
+// iterator returns a tree iterator in the specified direction.
+iterator :: proc "contextless" (t: ^$T/Tree($Key, $Value), direction: Direction) -> Iterator(Key, Value) {
+	it: Iterator(Key, Value)
+	it._tree      = transmute(^Tree(Key, Value))t
```
Comment (gingerBill) [inline core/container/rbtree/rbtree.odin:211]: A normal `cast` should do.

File: core/container/rbtree/rbtree.odin (hunk 1 above)
Comment (Kelimion) [reply core/container/rbtree/rbtree.odin:148]: I forgot to add `err: runtime.Allocator_Error` to the return values, like `core:container/avl` has, so I added that and updated the test to `_` the err.

File: core/container/rbtree/rbtree.odin (hunk 2 above)
Comment (Kelimion) [reply core/container/rbtree/rbtree.odin:211]: Done.

**Pull request merged**